### PR TITLE
LGA-3273 - Use esbuild to build and bundle assets

### PR DESCRIPTION
**Pull request by said-moj**
## What does this pull request do?

Use esbuild to build and bundle assets

## Any other changes that would benefit highlighting?

- Pinned node version to ^20.x, this is the same version required by the govuk-frontend library
- Updated pipeline to explicitly use the correct version of node
- Removed manual upgrade of `libexpat1` package as they latest changes are in the upstream repository
- Added a watcher service to local docker compose setup, it will rebuild assets when sass and js files in app/static/src are changed
- Added a`node_build` stage to the Dockerfile. This will build the assets which the production/development stages will copy from. This means that the final image does not need to have nodejs.

## Checklist

- [x] Provided JIRA ticket number in the title, e.g. "LGA-152: Sample title"

### CODE REVIEW

File: app/main/gtm.py [hunk 1]
```diff
@@ -32,7 +32,7 @@ def detect_gtm_anon_id():
     @after_this_request
     def remember_gtm_anon_id(response):
         session["gtm_anon_id"] = str(uuid.uuid4())
-        expiration_date = datetime.now(UTC) + timedelta(days=30)
+        expiration_date = datetime.utcnow() + timedelta(days=30)
```
Comment (BenMillar-MOJ) [inline app/main/gtm.py:35]: This function has been depreciated in favour of now(UTC)
[https://docs.python.org/3/library/datetime.html#datetime.datetime.utcnow](https://docs.python.org/3/library/datetime.html#datetime.datetime.utcnow)

**Review by BenMillar-MOJ — APPROVED** [review]
This is awesome, and works like a charm.
I have added a small note to the readme just specifying which node version we require.

File: app/static/src/js/custom.js [hunk 2]
```diff
@@ -1 +1 @@
-// Example of a custom js file. Which is combined with all JS files into one application JS file by Flask Assets.
```
Comment (said-moj) [inline app/static/src/js/custom.js:1]: Need to add the newline back

File: app/static/src/js/scripts.js [hunk 3]
```diff
@@ -0,0 +1,3 @@
+import './cookie-banner.js'
+import './gtm.js'
+import './custom.js'
```
Comment (said-moj) [inline app/static/src/js/scripts.js:3]: Have 2 scripts files, one for header and another for footer.

File: app/templates/base.html [hunk 4]
```diff
@@ -11,16 +11,10 @@
 {% block pageTitle %}{{config['SERVICE_NAME']}} – GOV.UK{% endblock %}
 
 {% block head %}
-  {% assets "headscripts" %}
-    <script nonce="{{ csp_nonce() }}">
-        gtm_anon_id = '{{ gtm_anon_id }}';
```
Comment (said-moj) [inline app/templates/base.html:16]: This needs go back in

File: app/templates/base.html [hunk 5]
```diff
@@ -170,12 +164,3 @@
     }
   }) }}
 {% endblock %}
-
-{% block bodyEnd %}
-    {% assets "js" %}
-        <script nonce="{{ csp_nonce() }}" type="module">
```
Comment (said-moj) [inline app/templates/base.html:176]: This needs to back in

File: esbuild.config.js [hunk 6]
```diff
@@ -0,0 +1,34 @@
+const esbuild = require('esbuild');
+const { sassPlugin } = require('esbuild-sass-plugin');
+
+
+const isProduction = process.env.NODE_ENV === 'production';
```
Comment (said-moj) [inline esbuild.config.js:5]: Use existing environment variables

File: esbuild.config.js [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+    const context = {
+        plugins: [sassPlugin({
+            quietDeps: true
+        })],
+        entryPoints: ["app/static/src/js/scripts.js", "app/static/src/scss/styles.scss"],
+        bundle: true,
+        entryNames: '[name]',
+        outdir: 'app/static/dist',   // Output directory,
+        minify: isProduction,
+        sourcemap: !isProduction,
```
Comment (said-moj) [inline esbuild.config.js:17]: Will be useful on production

File: .github/workflows/test.yml [hunk 8]
```diff
@@ -55,6 +50,11 @@ jobs:
       uses: actions/setup-python@v5
       with:
         python-version: '3.12'
+        cache: 'pip' # caching pip dependencies
+    - uses: actions/setup-node@v4
+      with:
+        node-version: '20.x'
+        cache: 'npm'
```
Comment (said-moj) [inline .github/workflows/test.yml:57]: Need to also `npm run build`

File: package.json [hunk 9]
```diff
@@ -1,11 +1,17 @@
...
+  "engines": {
+    "node": "^20.9.0",
+    "npm": "^10.1.0"
+  },
   "scripts": {
-    "gov-js": "cp -R node_modules/govuk-frontend/dist/govuk/govuk-frontend.min.* app/static",
-    "gov-images": "cp -R node_modules/govuk-frontend/dist/govuk/assets/. app/static",
-    "build": "npm run gov-js && npm run gov-images"
+    "gov-images": "cp -R node_modules/govuk-frontend/dist/govuk/assets/. app/static/dist",
+    "build": "npm run gov-images && node esbuild.config.js"
```
[the remark this answers is not in the document]
Comment (said-moj) [reply package.json]: Added

File: app/main/gtm.py (hunk 1 above)
Comment (said-moj) [reply app/main/gtm.py]: Updated to use `datetime.now(timezone.utc)`

**Pull request merged**